tokio/src/runtime/coop.rs

```diff
@@ -107,11 +105,98 @@ fn with_budget<R>(budget: Budget, f: impl FnOnce() -> R) -> R {
     f()
 }
 
-#[inline(always)]
-pub(crate) fn has_budget_remaining() -> bool {
-    // If the current budget cannot be accessed due to the thread-local being
-    // shutdown, then we assume there is budget remaining.
-    context::budget(|cell| cell.get().has_remaining()).unwrap_or(true)
+cfg_pub_if_rt! {
+    /// Returns `true` if there is still budget left on the task.
+    ///
+    /// Futures created by the tokio library functions are budget-aware and yield when there is no more
+    /// budget left, but not all futures will be budget-aware. Consider future `A` that polls two inner
+    /// futures `B` and `C`, and returns `Poll::Ready` when one of them is ready. If both inner futures
+    /// were budget-aware, at some point the budget would be depleted which would cause both futures to
+    /// return `Poll::Pending` resulting in `A` returning `Poll::Pending` as well. Yielding all the way
+    /// back to the runtime, the budget would be reset, and `B` and `C` could make progress again.
+    /// Now let's consider `B` is budget-aware, but `C` is not and is always ready. The budget will be
+    /// depleted as before, but now since `C` always returns `Poll::Ready`, `A` will always return
+    /// `Poll::Ready` as well. This way, `A` will not yield back to the runtime which will keep the budget
+    /// depleted and `B` not making progress.
```

Comment:
I don't like this macro. Is it possible to get rid of it? Possibly you could have a `pub(crate)` version that's always there and have the `pub` one call the `pub(crate)` one.